lib/clouds/as_cloud_ops.py

```diff
@@ -0,0 +1,151 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#/*******************************************************************************
+
+'''
+    Created on Jan 26, 2018
+    Azure Object Operations Library
+    @author: Marcio Silva, Michael R. Hines, Darrin Eden
+'''
+from time import time
+
+from lib.auxiliary.code_instrumentation import trace, cbdebug, cberr, cbwarn, cbinfo, cbcrit
+from lib.auxiliary.data_ops import is_number
+from libcloud_common import LibcloudCmds
+
+import hashlib
+
+#from libcloud.compute.drivers.azure import ConfigurationSet,ConfigurationSetInputEndpoint
+
+class AsCmds(LibcloudCmds) :
+    @trace
+    def __init__ (self, pid, osci, expid = None) :
+        LibcloudCmds.__init__(self, pid, osci, expid = expid, \
+                              provider = "AZURE", \
+                              num_credentials = 1, \
+                              use_ssh_keys = False, \
+                              use_volumes = False, \
+                              use_services = True, \
+                              use_get_image = False, \
+                              tldomain = "azure.com" \
+                             )
+    # All clouds based on libcloud should define this function.
+    # It performs the initial libcloud setup.
+    @trace
+    def get_libcloud_driver(self, libcloud_driver, tenant, access_token) :
+        '''
+        TBD
+        '''
+        _subscription_id, _certificate_path = access_token.split('+')
+        
+        driver = libcloud_driver(subscription_id = _subscription_id, key_file = _certificate_path)
+
+        return driver
+
+    @trace
+    def extra_vmc_setup(self, vmc_name, vmc_defaults, vm_defaults, vm_templates, _local_conn) :
+        '''
+        TBD
+        '''
+        
+        _cloud_service_found = False
+                
+        vm_defaults["cloud_service_name"] = vm_defaults["cloud_service_prefix"] + vmc_name.replace(' ','')
+        
+        vm_defaults["run_netname"] = "private"
+        
+        vmc_defaults["cloud_service_name"] = vm_defaults["cloud_service_name"]
+        
+        _cloud_service_name_list = _local_conn.ex_list_cloud_services()
+        
+        for _cloud_service in _cloud_service_name_list :
+            if _cloud_service.service_name == vm_defaults["cloud_service_name"] :
+                _cloud_service_found = True
+
+        if not _cloud_service_found :
+            _local_conn.ex_create_cloud_service(vm_defaults["cloud_service_name"], vmc_name)
+            LibcloudCmds.services = False
+            _cloud_service_found = True
+
+        hash_object = hashlib.sha1(str(time()))
+        hex_dig = hash_object.hexdigest()
+        vm_defaults["vm_name_suffix"] = str(hex_dig[0:10])
+
+        return _cloud_service_found
+
+    @trace
+    def get_list_node_args(self, obj_attr_list) :
```

Comment:
This `obj_attr_list["cloud_service_name"] ` is also Azure-specific. I advance here the same proposal as above (leave as is for now, and will fix it later).

Reply:
There's no sense of urgency here. The libcloud changes can be worked on more in their own branch/PR. Since nobody is explicitly using this adapter right now, we can try to get it right the first time.